### PR TITLE
[Doc] Describe the best scenarios to upgrade from v1 to v2

### PR DESCRIPTION
**Pull request by pierotibou**
Adding a document that aims at describing the rollout scenarios available for our customers when upgrading from v1 to v2

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by lucaspimentel** [comment]
> Trying to avoid triggering some OCD by switching to .NET everywhere

that would be me 😅

File: docs/migrations/v1_to_v2.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+For a more complete overview, please refer to our [release notes on GitHub](https://github.com/DataDog/dd-trace-dotnet/releases).
+
+## Upgrading from 1.x to 2.0
+
+### Code changes
+
+Most of our api changes do not require any changes to your code, but some patterns are no longer supported or recommended. Please refer to [Datadog.Trace documentation](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#upgrading-from-1x-to-20) for more information.
+
+### What if you are relying on .NET Framework lower than 4.6.0
```
Comment (lucaspimentel) [inline docs/migrations/v1_to_v2.md:19]:
```suggestion
### What if you are relying on .NET Framework lower than 4.6.1
```

File: docs/migrations/v1_to_v2.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Migration to .NET Tracer v2
+
+## .NET Tracer v2.0 content
+
+The .NET Tracer v2.0:
+
+- Fixes a long standing bug where traces could be disconnected when automatic and custom tracing were using two different versions of the tracer.
+- Adds support to .NET 6 and drops the support of .NET Framework 4.5 to .NET Framework 4.6.1. If you are currently targeting version < 4.6.1, we suggest you upgrade in line with Microsoft's guidance.
```
Comment (lucaspimentel) [inline docs/migrations/v1_to_v2.md:8]:
```suggestion
- Adds support for .NET 6 and ends support of .NET Framework versions older than 4.6.1. If you are currently targeting version < 4.6.1, we suggest you upgrade in line with Microsoft's guidance.
```

File: docs/migrations/v1_to_v2.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+- .NET Framework 4.6.1 or above
+- .NET Core 3.1 or above
+- .NET Standard 2.0 or above (includes .NET Core 2.0 and above)
+
+### Rollout strategy when mixing automatic and custom instrumentation
+
+**Note that** even though you won't get disconnected traces when using two different v2 versions of the tracer, **you should aim at aligning the versions of the automatic and manual tracers**. Indeed, there's a bigger overhead when there's a version mismatch, so you should aim at making this situation temporary.
+
+#### .NET Core
```
Comment (lucaspimentel) [inline docs/migrations/v1_to_v2.md:35]: Adding .NET 5+ (to be clear about .NET Core vs .NET 5/6)
```suggestion
#### .NET Core and .NET 5+
```

File: docs/migrations/v1_to_v2.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+### What if you are relying on .NET Framework lower than 4.6.0
+
+If you are currently targeting version < 4.6.1, we suggest you upgrade in line with [Microsoft's guidance](https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-framework). On our end, we will support fixing major bugs on top of Tracer v1.31.
+
+That said, automatic instrumentation relies only on the CLR (runtime) version, not the compile-time target. Even if an application is built for .NET Framework older than 4.6.1, as long as it runs on a supported CLR version, automatic instrumentation will work normally.
+
+Regarding custom instrumentation, NuGet package version 2.0 and above will no longer include binaries for versions of .NET Framework older than 4.6.1. The NuGet package will only work correctly with applications that target (at compile time):
+
+- .NET Framework 4.6.1 or above
+- .NET Core 3.1 or above
```
Comment (lucaspimentel) [inline docs/migrations/v1_to_v2.md:28]: Adding .NET 5+ (to be clear about .NET Core vs .NET 5/6)
```suggestion
- .NET Core 3.1 or above
- .NET 5 or above
```

File: docs/migrations/v1_to_v2.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+Indeed, on .NET Core, if the assembly version from the nuget package is newer, the CLR can load that one instead of the assembly from the tracer's home folder.
+
+##### If you were using v1.28.7 or lower
+
+In that case, the migration will create disconnected traces one last time. **You should update the Tracer Home (ie serverside) first though**, then the nuget.
+
+#### .NET Framework
+
+If you rely on the .NET Framework, there are no scenarios where you could avoid disconnected traces between automatic and custome instrumentation. In that case, you should do what's easier for you.
```
Comment (lucaspimentel) [inline docs/migrations/v1_to_v2.md:52]:
```suggestion
If you rely on the .NET Framework, there are no scenarios where you could avoid disconnected traces between automatic and custom instrumentation. In that case, you should do what's easier for you.
```

**Review by lucaspimentel — COMMENTED** [review]
A few tweaks to supports versions

File: docs/migrations/v1_to_v2.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+That said, automatic instrumentation relies only on the CLR (runtime) version, not the compile-time target. Even if an application is built for .NET Framework older than 4.6.1, as long as it runs on a supported CLR version, automatic instrumentation will work normally.
+
+Regarding custom instrumentation, NuGet package version 2.0 and above will no longer include binaries for versions of .NET Framework older than 4.6.1. The NuGet package will only work correctly with applications that target (at compile time):
+
+- .NET Framework 4.6.1 or above
+- .NET Core 3.1 or above
+- .NET 5 or above
+- .NET Standard 2.0 or above (includes .NET Core 2.0 and above)
+
+### Rollout strategy when mixing automatic and custom instrumentation
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:32]:
```suggestion
### Rollout strategy when using automatic and custom instrumentation
```

File: docs/migrations/v1_to_v2.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# Migration to .NET Tracer v2
+
+## .NET Tracer v2.0 content
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:3]:
```suggestion
## .NET Tracer v2.0 contents
```

File: docs/migrations/v1_to_v2.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+Regarding custom instrumentation, NuGet package version 2.0 and above will no longer include binaries for versions of .NET Framework older than 4.6.1. The NuGet package will only work correctly with applications that target (at compile time):
+
+- .NET Framework 4.6.1 or above
+- .NET Core 3.1 or above
+- .NET 5 or above
+- .NET Standard 2.0 or above (includes .NET Core 2.0 and above)
+
+### Rollout strategy when mixing automatic and custom instrumentation
+
+**Note that** even though you won't get disconnected traces when using two different v2 versions of the tracer, **you should aim at aligning the versions of the automatic and manual tracers**. Indeed, there's a bigger overhead when there's a version mismatch, so you should aim at making this situation temporary.
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:34]: Several suggestions wrapped into one. We can workshop this, but here's my initial rewording:

```suggestion
If your application uses both automatic and custom instrumentation, your application traces may be disconnected until both components are upgraded to v2.x. Follow the steps below to upgrade the components in the recommended order.

**Note:** To minimize application overhead, align the versions of the automatic and custom instrumentation. Using two different versions of the 2.x .NET Tracer will still result in complete traces but it requires additional overhead, so Datadog recommends minimizing the amount of time that the versions are mismatched.
```

File: docs/migrations/v1_to_v2.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+#### .NET Core and .NET 5+
+
+##### If you were using v1.28.8 or greater
+
+On .NET Core, to avoid disconnected traces during the upgrade process:
+
+- Upgrade the nuget package to the latest version of the tracer
+- Then upgrade serverside
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:43]:
```suggestion
1. Upgrade the nuget package to the latest version of the tracer
2. Upgrade the automatic instrumentation package to the matching version of the tracer
```

File: docs/migrations/v1_to_v2.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+Indeed, on .NET Core, if the assembly version from the nuget package is newer, the CLR can load that one instead of the assembly from the tracer's home folder.
+
+##### If you were using v1.28.7 or lower
+
+In that case, the migration will create disconnected traces one last time. **You should update the Tracer Home (ie serverside) first though**, then the nuget.
+
+#### .NET Framework
+
+If you rely on the .NET Framework, there are no scenarios where you could avoid disconnected traces between automatic and custom instrumentation. In that case, you should do what's easier for you.
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:53]:
```suggestion
On .NET Framework there will be disconnected traces between automatic and custom instrumentation until both packages are upgraded to 2.x. In this scenario, Datadog recommends upgrading the automatic and custom instrumentation in whichever order is easiest for your application.
```

File: docs/migrations/v1_to_v2.md (hunk 8 above)
Comment (pierotibou) [reply docs/migrations/v1_to_v2.md:34]: It looks good IMO. Committing your suggestion as is.
Thanks.

File: docs/migrations/v1_to_v2.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+- Adds support for .NET 6 and ends support of .NET Framework versions older than 4.6.1. If you are currently targeting version < 4.6.1, we suggest you upgrade in line with Microsoft's guidance.
+- Refactored our APIs for an easier and safer use.
+
+For a more complete overview, please refer to our [release notes on GitHub](https://github.com/DataDog/dd-trace-dotnet/releases).
+
+## Upgrading from 1.x to 2.0
+
+### Code changes
+
+Most of our api changes do not require any changes to your code, but some patterns are no longer supported or recommended. Please refer to [Datadog.Trace documentation](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#upgrading-from-1x-to-20) for more information.
```
Comment (andrewlock) [inline docs/migrations/v1_to_v2.md:17]: Knowing that people don't read text, and even less so links, I'd be inclined to outline the bullet point headings here, and just link to the main doc

```suggestion
Most of our api changes do not require any changes to your code, but some patterns are no longer supported or recommended. Please refer to [Datadog.Trace documentation](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#upgrading-from-1x-to-20) for full details:
- [Singleton `Tracer` instances](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#supported-net-versions)
- [Immutable `Tracer.Settings`](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#immutable-tracersettings)
- [Exporter settings](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#exporter-settings)
- [Configure ADO.NET integrations individually](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#configure-adonet-integrations-individually)
- [`ElasticsearchNet5` integration ID removed](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#elasticsearchnet5-integration-id-removed)
- [Obsolete APIs have been removed](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#obsolete-apis-have-been-removed)
- [Introduction of interfaces `ISpan`, `IScope`, and `ITracer`](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#elasticsearchnet5-integration-id-removed)
- [Simplification of the tracer interface](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#simplification-of-the-tracer-interface)
- [Incorrect integration names are ignored](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#incorrect-integration-names-are-ignored)
- [Automatic instrumentation changes](https://github.com/DataDog/dd-trace-dotnet/tree/v2.0.1/docs/Datadog.Trace#automatic-instrumentation-changes)
```

File: docs/migrations/v1_to_v2.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+**Note:** To minimize application overhead, align the versions of the automatic and custom instrumentation. Using two different versions of the 2.x .NET Tracer will still result in complete traces but it requires additional overhead, so Datadog recommends minimizing the amount of time that the versions are mismatched.
+
+#### .NET Core and .NET 5+
+
+##### If you were using v1.28.8 or greater
+
+On .NET Core, to avoid disconnected traces during the upgrade process:
+
+1. Upgrade the nuget package to the latest version of the tracer
```
Comment (andrewlock) [inline docs/migrations/v1_to_v2.md:44]:
```suggestion
1. Upgrade the NuGet package to the latest version of the tracer
```

File: docs/migrations/v1_to_v2.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+2. Upgrade the automatic instrumentation package to the matching version of the tracer
+
+Indeed, on .NET Core, if the assembly version from the nuget package is newer, the CLR can load that one instead of the assembly from the tracer's home folder.
```
Comment (andrewlock) [inline docs/migrations/v1_to_v2.md:47]: I would rephrase this a little:
```suggestion
> If you are targeting .NET Core/.NET 5+, and are unable to upgrade both the _Datadog.Trace_ NuGet package and MSI simultaneously, always prefer upgrading the NuGet package first.
```

File: docs/migrations/v1_to_v2.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+##### If you were using v1.28.7 or lower
+
+In that case, the migration will create disconnected traces one last time. **You should update the Tracer Home (ie serverside) first though**, then the nuget.
```
Comment (andrewlock) [inline docs/migrations/v1_to_v2.md:51]: Really? 🤔 So for <1.28.7 they should do MSI then NuGet, but for >=1.28.8, they do NuGet first? What's the reason for that? Did we relax the netcore loading conditions in 1.28.8? What happens if they do NuGet first in that situation?

```suggestion
In that case, the migration will create disconnected traces one last time. **You should update the Tracer Home (ie serverside) first though**, then the NuGet.
```

File: docs/migrations/v1_to_v2.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+#### .NET Framework
+
+On .NET Framework there will be disconnected traces between automatic and custom instrumentation until both packages are upgraded to 2.x. In this upgrade scenario, Datadog recommends upgrading the automatic and custom instrumentation in whichever order is easiest for your application.
```
Comment (andrewlock) [inline docs/migrations/v1_to_v2.md:55]:
```suggestion
On .NET Framework there will be disconnected traces between automatic and custom instrumentation until both packages are upgraded to 2.x. In this upgrade scenario, Datadog recommends upgrading the MSI and NuGet package in whichever order is easiest for your application.
```

**Review by andrewlock — COMMENTED** [review]
I have a couple of small suggestions/questions

File: docs/migrations/v1_to_v2.md (hunk 14 above)
Comment (pierotibou) [reply docs/migrations/v1_to_v2.md:51]: I don't recall the details we had discussed at the time. The conclusion was that upgrading the NuGet first would lead to missing traces as IIRC there were breaking changes before that (CallTarget vs CallSite maybe, sorry I really don't remember). 
So upgrading the NuGet first would make us enter that scenario because of .NetCore loading conditions. Thus why we suggest starting with the MSI as disconnected traces was better that missing ones.

cc @lucaspimentel and @kevingosse cause the only thing I recall is that we have discussed that together :)

Comment (lucaspimentel) [reply docs/migrations/v1_to_v2.md]: > the only thing I recall is that we have discussed that together

Then you remember more than I do 😅

Comment (pierotibou) [reply docs/migrations/v1_to_v2.md]: Ok I have done a bit of archeology in the comments of the file. @lucaspimentel mentionned that previous to version 1.28.6 we could face crashed as - in this version - he has merged multiple DLLs into one to avoid making functions calls across assemblies that could potentially have internal breaking changes. [https://github.com/DataDog/dd-trace-dotnet/pull/1443](https://www.google.com/url?q=https://github.com/DataDog/dd-trace-dotnet/pull/1443&sa=D&source=docs&ust=1644836181552487&usg=AOvVaw0U9HIz4Vs82FDcc3yLZCBp)

File: docs/migrations/v1_to_v2.md [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+Regarding custom instrumentation, NuGet package version 2.0 and above will no longer include binaries for versions of .NET Framework older than 4.6.1. The NuGet package will only work correctly with applications that target (at compile time):
+
+- .NET Framework 4.6.1 or above
+- .NET Core 3.1 or above
+- .NET 5 or above
+- .NET Standard 2.0 or above (includes .NET Core 2.0 and above)
+
+### Rollout strategy when using automatic and custom instrumentation
+
+If your application uses both automatic and custom instrumentation, your application traces may be disconnected until both components are upgraded to v2.x. Follow the steps below to upgrade the components in the recommended order.
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:44]:
```suggestion
If your application uses both automatic and custom instrumentation, your application traces may be disconnected until both components are upgraded to v2.x. If you are unable to upgrade both components simultaneously, follow the steps below to upgrade the components in the recommended order.
```

File: docs/migrations/v1_to_v2.md [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+#### .NET Core and .NET 5+
+
+##### If you were using v1.28.8 or greater
+
+On .NET Core, to avoid disconnected traces during the upgrade process:
+
+1. Upgrade the NuGet package to the latest version of the tracer
+2. Upgrade the automatic instrumentation package to the matching version of the tracer
+
+> If you are targeting .NET Core/.NET 5+, and are unable to upgrade both the _Datadog.Trace_ NuGet package and MSI simultaneously, always prefer upgrading the NuGet package first.
```
Comment (zacharycmontoya) [inline docs/migrations/v1_to_v2.md:57]: I added a comment in the "rollout strategy" section above that could make this comment redundant. What do you think of the suggestion to move it up?

File: docs/migrations/v1_to_v2.md (hunk 14 above)
Comment (pierotibou) [reply docs/migrations/v1_to_v2.md]: And to close the loop on this one, I've done the test that I should have done when the topic has been raised and indeed with a `1.28.4` msi and a 2.x nuget, we lose automatic spans, we just get custom ones (without any error in logs btw which seems weird).

File: docs/migrations/v1_to_v2.md (hunk 17 above)
Comment (pierotibou) [reply docs/migrations/v1_to_v2.md:57]: 🤔, I don't see exactly why. I think this comment works only in v1.28.8 and above section. But we can discuss offline and I can rearrange obv.

Comment (pierotibou) [reply docs/migrations/v1_to_v2.md]: Discussed offline. Thanks

**Review by lucaspimentel — APPROVED** [review]

File: docs/migrations/v1_to_v2.md (hunk 14 above)
Comment (zacharycmontoya) [reply docs/migrations/v1_to_v2.md]: > we lose automatic spans, we just get custom ones (without any error in logs btw which seems weird).

Interesting... I definitely expected logs because the 2.x Datadog.Trace will get loaded at startup and then PInvoke back to the profiler to load all of the integrations, and at this point the PInvoke should fail because the older `1.28.4` doesn't have the methods
